### ISSUE TITLE
How to enforce a specific composition of participants?

### ISSUE THREAD

**Issue by eamanj** [empiricaly/meteor-empirica-core#55]
I have a game in which participants come from a panel with known attributes. We direct them to the platform with their attributes in the link. They still have the chance to update their attributes in an intro stage survey. How can I enforce a specific composition of subjects based on an attribute X? For example, I want 3 players with value of x and 2 players with value of y in their attribute X in the game stage. Is there any way to do this with the platform? If it requires making small changes in the core, can you direct me how to do this?

**Comment by npaton** [comment]
There is currently no easy way to do this. Assignment to a game happens early, and it is difficult to change the current system's behavior.

We have recognized this issue and we are working on a large change to Empirica that will address this case and other limitations we have come across. This new release of Empirica is not planned for the near future, but should be available later this year, hopefully this summer.

**Comment by joshua-a-becker** [comment]
@eamanj  ----  you can hack a solution to this by using your game.init() function to assign people to groups.   It's counter to Empirica's design, but gets the job done.   The downside is that it's "leaky" for participants---i.e., you may find yourself unable to use all participants, and maybe have to pay them something just for showing up.

@npaton  @amaatouq --- i've though a lot about game assignment... i don't have any perfect solution (i'm sure you've already faced a lot of the inevitable tradeoffs) and i'd love to join the brainstorm if it's still open for redesign.  i'm currently sitting here looking at my server using 0.3 ECU just to sit idle, because I have a batch with 100 single-person game lobbies running.

**Comment by npaton** [comment]
@joshua-a-becker Let's plan to meet soon to discuss this, we'd love to get your input.

**Comment by eamanj** [comment]
@joshua-a-becker are you saying that I should ask for more participants than I need and hope that the right composition eventually arrives? and send the redundant subjects to exit survey?

**Comment by amaatouq** [comment]
No, what I believe josh meant is to have playerCount be 20 for example and then create at gameInit() 5 connected components (5 networks) of 4 players each.. in gameinit() you have access to their affiliation so you can decide to make each network exactly 2 and 2... then you have to change the code throughout such that people only observe their peers, not everyone in game.players list.

**Comment by joshua-a-becker** [comment]
you're both right --- @amaatouq is describing the technique, but this does necessarily result in overflow, so you might have to send some players away.  like, if you need each group to have 50/50 democrats/republicans but you start with a 70/30 composition.... then you end up discarding 40% of your players.  

**AND** i was imagining this solution, i realize, in the paradigm where everybody starts at the exact same time.  it's more challenging for the trickle-in approach i'd guess.   happy to share the tools i use for simultaneous start, which brings a host of other challenges.  

admittedly a far from perfect solution---i was just brainstorming.   could be useful for a pilot test but it gets expensive in the long run.

**Comment by joshua-a-becker** [comment]
though----i feel like you'd have to discard 40% of your players no matter what the method?  Doug can probably comment better on that as he has experience with this.

**Comment by amaatouq** [comment]
Or have your treatment be more continues (ie the level of polarization/diversity; rather than a binary treatement) and creating groups at random will allow you to make your causal claim ..  also, you can make sure to qualify 50% of each type (even if the population is unbalanced, ie 70% 30%, which means you’ll over sample the minority group), which will increase your statsitical power.
